src/analyses/mutexGhosts.ml

```diff
@@ -0,0 +1,244 @@
+(** Analysis for generating ghost variables corresponding to mutexes ([mutexGhosts]). *)
+
+open Analyses
+
+module NodeSet = Queries.NS
+
+
+module Spec =
+struct
+  include UnitAnalysis.Spec
+  let name () = "mutexGhosts"
+
+  (* module ThreadCreate = Printable.UnitConf (struct let name = "threadcreate" end) *)
+  module V =
+  struct
+    include Printable.Either3 (Node) (LockDomain.MustLock) (BoolDomain.Bool)
+    let node x = `Left x
+    let lock x = `Middle x
+    let threadcreate = `Right false
+    let update = `Right true
+    let is_write_only = function
+      | `Left _ -> false
+      | `Middle _ -> true
+      | `Right false -> false
+      | `Right true -> true
+  end
+
+  module Locked =
+  struct
+    include LockDomain.MayLocksetNoRW
+    let name () = "locked"
+  end
+  module Unlocked =
+  struct
+    include LockDomain.MayLocksetNoRW
+    let name () = "unlocked"
+  end
+  module MultiThread =
+  struct
+    include BoolDomain.MayBool
+    let name () = "multithread"
+  end
+  module G =
+  struct
+    include Lattice.Lift2 (Lattice.Prod3 (Locked) (Unlocked) (MultiThread)) (Lattice.Lift2 (BoolDomain.MayBool) (NodeSet))
+    let node = function
+      | `Bot -> (Locked.bot (), Unlocked.bot (), MultiThread.bot ())
+      | `Lifted1 x -> x
+      | _ -> failwith "MutexGhosts.node"
+    let lock = function
+      | `Bot -> BoolDomain.MayBool.bot ()
+      | `Lifted2 (`Lifted1 x) -> x
+      | _ -> failwith "MutexGhosts.lock"
+    let threadcreate = function
+      | `Bot -> NodeSet.bot ()
+      | `Lifted2 (`Lifted2 x) -> x
+      | _ -> failwith "MutexGhosts.threadcreate"
+    let update = threadcreate
+    let create_node node = `Lifted1 node
+    let create_lock lock = `Lifted2 (`Lifted1 lock)
+    let create_threadcreate threadcreate = `Lifted2 (`Lifted2 threadcreate)
+    let create_update = create_threadcreate
+  end
+
+  let mustlock_of_addr (addr: LockDomain.Addr.t): LockDomain.MustLock.t option =
+    match addr with
+    | Addr mv when LockDomain.Mval.is_definite mv -> Some (LockDomain.MustLock.of_mval mv)
+    | _ -> None
+
+  let event ctx e octx =
+    let verifier_atomic_addr = LockDomain.Addr.of_var LibraryFunctions.verifier_atomic_var in
+    begin match e with
+      | Events.Lock (l, _) when not (LockDomain.Addr.equal l verifier_atomic_addr) ->
+        ctx.sideg (V.node ctx.prev_node) (G.create_node (Locked.singleton l, Unlocked.bot (), MultiThread.bot ()));
```

Comment:
Does this side-effect need to be made during the analysis?

Reply:
I think yes, because the following logic checks if the set is non-singleton, but that needs to account for side effects to the same unknown with other values that could potentially make it non-singleton. The [V.node] unknowns are not write-only because they are read below to determine values for [V.update] unknowns.